core/shared/src/main/scala/zio/config/experimental/package.scala

```diff
@@ -0,0 +1,186 @@
+package zio.config
+
+import zio._
+
+package object experimental {
+
+  def fromIndexedFlat(indexedFlat: IndexedFlat): ConfigProvider =
+    new ConfigProvider {
+
+      import Config._
+      import IndexedFlat._
+
+      def extend[A, B](leftDef: Int => A, rightDef: Int => B)(left: Chunk[A], right: Chunk[B]): (Chunk[A], Chunk[B]) = {
+        val leftPad = Chunk.unfold(left.length) { index =>
+          if (index >= right.length) None else Some(leftDef(index) -> (index + 1))
+        }
+        val rightPad = Chunk.unfold(right.length) { index =>
+          if (index >= left.length) None else Some(rightDef(index) -> (index + 1))
+        }
+
+        val leftExtension  = left ++ leftPad
+        val rightExtension = right ++ rightPad
+
+        (leftExtension, rightExtension)
+      }
+
+      def loop[A](prefix: IndexedFlat.ConfigPath, config: Config[A], split: Boolean)(implicit
+        trace: Trace
+      ): IO[Config.Error, Chunk[A]] =
+        config match {
+          case fallback: Fallback[A] =>
+            loop(prefix, fallback.first, split).catchAll(e1 =>
+              if (fallback.condition(e1)) loop(prefix, fallback.second, split).catchAll(e2 => ZIO.fail(e1 || e2))
+              else ZIO.fail(e1)
+            )
+
+          case Described(config, _) => loop(prefix, config, split)
+
+          case Lazy(thunk) => loop(prefix, thunk(), split)
+
+          case MapOrFail(original, f) =>
+            loop(prefix, original, split).flatMap { as =>
+              ZIO.foreach(as)(a => ZIO.fromEither(f(a)).mapError(_.prefixed(prefix.map(_.value))))
+            }
+
+          case Sequence(config) =>
+            for {
+              keys <- indexedFlat
+                        .enumerateChildrenIndexed(prefix)
+                        .map(set =>
+                          set.toList.flatMap { chunk =>
+                            chunk.headOption.toList
+                          }
```

Comment:
@adamgfraser  As discussed, copied over to zio-config.
https://github.com/zio/zio-config/pull/1035/commits/b4b225971e0973886849f1106a9737e3172c2e84

I think, we have still got a failing test if someone configures for `Option[List]`. Although it might sound redundant, it implies, you may skip the key as such, so that we get a None. 

The above implementation returns `Some(Nil)`, when the key is absent. I think good to keep a test if merging this PR. Or else, I can take care of it in zio-config

Reply:
The solution is before enumerating the prefix, we need to search the map if any key starts with `prefix`

Reply:
@afsalthaj If it is an issue with an "indexed" flat configuration provider let's take it up in ZIO Config. If it is an issue with a traditional flat configuration provider can you create a small reproducer and I will take care of it here or in a follow up?